near-sdk-macros/src/core_impl/code_generator/serializer.rs

```diff
@@ -0,0 +1,30 @@
+use crate::core_impl::info_extractor::{AttrSigInfo, SerializerType};
+use proc_macro2::TokenStream as TokenStream2;
+use quote::quote;
+
+pub fn generate_serializer(
+    attr_sig_info: &AttrSigInfo,
+    serializer: &SerializerType,
+) -> TokenStream2 {
+    let has_input_args = attr_sig_info.input_args().next().is_some();
+    if !has_input_args {
+        return quote! { let args = vec![]; };
+    }
+    let struct_decl = attr_sig_info.input_struct_ser();
+    let constructor_call = attr_sig_info.constructor_expr_ref();
+    let constructor = quote! { let args = #constructor_call; };
+    let value_ser = match serializer {
+        SerializerType::JSON => quote! {
+            let args = near_sdk::serde_json::to_vec(&args).expect("Failed to serialize the cross contract args using JSON.");
+        },
+        SerializerType::Borsh => quote! {
+            let args = near_sdk::borsh::BorshSerialize::try_to_vec(&args).expect("Failed to serialize the cross contract args using Borsh.");
+        },
+    };
+
+    quote! {
+      #struct_decl
+      #constructor
+      #value_ser
+    }
```

Comment:
Yeah, this is somewhat expected since the syntax within macros isn't necessarily Rust. There are things like https://github.com/dtolnay/prettyplease which I think moves in that direction, but probably hard to have for the general case.